core/trino-main/src/main/java/io/trino/execution/scheduler/policy/PhasedExecutionSchedule.java

```diff
@@ -410,19 +412,18 @@ private FragmentSubGraph processJoin(boolean replicated, PlanNode probe, PlanNod
             addDependencyEdges(buildSubGraph.getUpstreamFragments(), probeSubGraph.getLazyUpstreamFragments());
 
             boolean currentFragmentLazy = probeSubGraph.isCurrentFragmentLazy() && buildSubGraph.isCurrentFragmentLazy();
-            if (replicated && currentFragmentLazy) {
-                // Do not start join stage (which can also be a source stage with table scans)
-                // for replicated join until build source stage enters FLUSHING state.
-                // Broadcast join limit for CBO is set in such a way that build source data should
-                // fit into task output buffer.
-                // In case build source stage is blocked on full task buffer then join stage
-                // will be started automatically regardless od dependency. This is handled by
-                // unblockStagesWithFullOutputBuffer method.
+            if (fragments.get(currentFragmentId).getPartitioning().equals(SOURCE_DISTRIBUTION) && currentFragmentLazy) {
```

Comment:
From what I understand it is to prevent a non source distributed stage running a replicated join from being executed lazily. Is it correct? I wonder why is it necessary.